// @aws-cdk/aws-dynamodb/lib/replica-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReplicaProvider = void 0;
const path = require("path");
const iam = require("@aws-cdk/aws-iam");
const lambda = require("@aws-cdk/aws-lambda");
const core_1 = require("@aws-cdk/core");
const cr = require("@aws-cdk/custom-resources");
class ReplicaProvider extends core_1.NestedStack {
    constructor(scope, id, props = {}) {
        super(scope, id);
        const code = lambda.Code.fromAsset(path.join(__dirname, 'replica-handler'));
        // Issues UpdateTable API calls
        this.onEventHandler = new lambda.Function(this, 'OnEventHandler', {
            code,
            runtime: lambda.Runtime.NODEJS_12_X,
            handler: 'index.onEventHandler',
            timeout: core_1.Duration.minutes(5),
        });
        // Checks if table is back to `ACTIVE` state
        this.isCompleteHandler = new lambda.Function(this, 'IsCompleteHandler', {
            code,
            runtime: lambda.Runtime.NODEJS_12_X,
            handler: 'index.isCompleteHandler',
            timeout: core_1.Duration.seconds(30),
        });
        // Allows the creation of the `AWSServiceRoleForDynamoDBReplication` service linked role
        this.onEventHandler.addToRolePolicy(new iam.PolicyStatement({
            actions: ['iam:CreateServiceLinkedRole'],
            resources: [core_1.Stack.of(this).formatArn({
                    service: 'iam',
                    region: '',
                    resource: 'role',
                    resourceName: 'aws-service-role/replication.dynamodb.amazonaws.com/AWSServiceRoleForDynamoDBReplication',
                })],
        }));
        // Required for replica table creation
        this.onEventHandler.addToRolePolicy(new iam.PolicyStatement({
            actions: ['dynamodb:DescribeLimits'],
            resources: ['*'],
        }));
        this.provider = new cr.Provider(this, 'Provider', {
            onEventHandler: this.onEventHandler,
            isCompleteHandler: this.isCompleteHandler,
            queryInterval: core_1.Duration.seconds(10),
            totalTimeout: props.timeout,
        });
    }
    /**
     * Creates a stack-singleton resource provider nested stack.
     */
    static getOrCreate(scope, props = {}) {
        var _a;
        const stack = core_1.Stack.of(scope);
        this.checkManagedPoliciesLimit(stack);
        const uid = '@aws-cdk/aws-dynamodb.ReplicaProvider';
        return (_a = stack.node.tryFindChild(uid)) !== null && _a !== void 0 ? _a : new ReplicaProvider(stack, uid, props);
    }
    static checkManagedPoliciesLimit(stack) {
        var _a;
        // The custom resource implementation uses IAM managed policies. There's
        // a limit of 10 managed policies per role in IAM. Throw if we reach this
        // limit.
        const calls = (_a = this.getOrCreateCalls.get(stack.stackName)) !== null && _a !== void 0 ? _a : 0;
        if (calls >= 10) {
            throw new Error('You currently cannot have more than 10 global tables in a single stack. Consider splitting your tables across multiple stacks.');
        }
        this.getOrCreateCalls.set(stack.stackName, calls + 1);
    }
}
exports.ReplicaProvider = ReplicaProvider;
// Map of getOrCreate() calls per stack
ReplicaProvider.getOrCreateCalls = new Map();
